pkg/pool-stable/contracts/meta/StableOracleMath.sol

```diff
@@ -0,0 +1,118 @@
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/LogCompression.sol";
+
+import "../StableMath.sol";
+import "hardhat/console.sol";
+
+contract StableOracleMath is StableMath {
+    using FixedPoint for uint256;
+
+    /**
+     * @dev Calculates the spot price of token Y and BPT in token X.
+     */
+    function _calcLogPrices(
+        uint256 amplificationParameter,
+        uint256 balanceX,
+        uint256 balanceY,
+        int256 logBptTotalSupply
+    ) internal pure returns (int256 logSpotPrice, int256 logBptPrice) {
+        uint256 spotPrice = _calcSpotPrice(amplificationParameter, balanceX, balanceY);
+        logBptPrice = _calcLogBptPrice(spotPrice, balanceX, balanceY, logBptTotalSupply);
+        logSpotPrice = LogCompression.toLowResLog(spotPrice);
+    }
+
+    /**
+     * @dev Calculates the spot price of token Y in token X.
+     */
+    function _calcSpotPrice(
+        uint256 amplificationParameter,
+        uint256 balanceX,
+        uint256 balanceY
+    ) internal pure returns (uint256) {
+        /**************************************************************************************************************
+        //                                                                                                           //
+        //                             2.a.x.y + a.y^2 + b.y                                                         //
+        // spot price Y/X = - dx/dy = -----------------------                                                        //
+        //                             2.a.x.y + a.x^2 + b.x                                                         //
+        //                                                                                                           //
+        // n = 2                                                                                                     //
+        // a = amp param * n                                                                                         //
+        // b = D + a.(S - D)                                                                                         //
+        // D = invariant                                                                                             //
+        // S = sum of balances but x,y = 0                                                                           //
+        **************************************************************************************************************/
+
+        uint256 invariant = _calculateInvariant(amplificationParameter, _balances(balanceX, balanceY), true);
+
+        uint256 a = (amplificationParameter * 2) / _AMP_PRECISION;
+        uint256 b = Math.mul(invariant, a).sub(invariant);
+
+        uint256 axy2 = Math.mul(a * 2, balanceX).mulUp(balanceY);
+
+        // dx = a.x.y.2 + a.y^2 - b.y
+        uint256 derivativeX = axy2.add(Math.mul(a, balanceY).mulUp(balanceY)).sub(b.mulUp(balanceY));
```

Comment:
`mulDown` is slightly less expensive than `mulUp`, and we don't care about rounding direction (since we introduce a lot of error later on).